com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs

```diff
@@ -99,15 +95,78 @@ static string[] GatherDependenciesFromSourceFile(string assetPath)
             }
         }
 
-        public override void OnImportAsset(AssetImportContext ctx)
+        Shader BuildAllShaders(
+            AssetImportContext importContext,
+            AssetImportErrorLog importErrorLog,
+            AssetCollection allImportAssetDependencies,
+            GraphData graph)
         {
-            var oldShader = AssetDatabase.LoadAssetAtPath<Shader>(ctx.assetPath);
-            if (oldShader != null)
-                ShaderUtil.ClearShaderMessages(oldShader);
+            Shader primaryShader = null;
 
-            var importLog = new AssetImportErrorLog(ctx);
+            string path = importContext.assetPath;
+            var primaryShaderName = Path.GetFileNameWithoutExtension(path);
+
+            try
+            {
+                // this will also add Target dependencies into the asset collection
+                Generator generator;
+                generator = new Generator(graph, graph.outputNode, GenerationMode.ForReals, primaryShaderName, assetCollection: allImportAssetDependencies);
+
+                bool first = true;
+                foreach (var generatedShader in generator.allGeneratedShaders)
+                {
+                    var shaderString = generatedShader.codeString;
+
+                    // we only care if an error was reported for a node that we actually used
+                    if (graph.messageManager.AnyError((nodeId) => NodeWasUsedByGraph(nodeId, graph)) ||
+                        shaderString == null)
+                    {
+                        shaderString = k_ErrorShader.Replace("Hidden/GraphErrorShader2", generatedShader.shaderName);
+                    }
+
+                    var shader = ShaderUtil.CreateShaderAsset(importContext, generatedShader.codeString, false);
```

Comment:
nit: `generatedShader.codeString` -> `shaderString`

Reply:
Actually, I assume this should use the error shader if that's encountered too, so maybe not a nit?

Reply:
yes, this should be shaderString